### PR TITLE
Contributor's Guide for Durable JS

### PR DESCRIPTION
**Pull request by priyaananthasankar**
https://github.com/Azure/azure-functions-durable-js/issues/123

### CODE REVIEW

**Comment by priyaananthasankar** [comment]
cc: @ConnorMcMahon : Can you please review this?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+The following extensions should be installed if using Visual Studio Code for debugging:
+
+- C# for Visual Studio Code (powered by OmniSharp)
+- Azure Functions Extensions for Visual Studio Code v0.19.1 and above.
+
+### Setting up durable-js debugging
+
+- Create a Durable Functions Orchestrator for FunctionChaining pattern using [starter templates](https://docs.microsoft.com/en-us/azure/azure-functions/durable/quickstart-js-vscode)
+  Note: In this starter template, ignore the line that says: "On a Mac or Linux computer, you must set the AzureWebJobsStorage property to the connection string of an existing Azure storage account". We will be setting up the AzureWebJobsStorage property to `UseDevelopmentStorage=true`
```
Comment (ConnorMcMahon) [inline CONTRIBUTING.md]: If you found that Azurite is sufficient for local dev, it may be a good idea for us to update this line in our template to give that option.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+### Debugging end-to-end
+
+1. Open the Azure Storage Explorer and connect to the local storage emulator or the storage account you are using.
+2. In the VSCode editor for durable-js click Debug -> Start Debugging. This will internally start `func host start` through core tools and provides the orchestrator client URL
+3. In the VSCode editor for DurableTask, click Debug -> .NET Core Attach Process and search for `func host start` process and attach to it.
+4. Add a breakpoint in both editors and continue debugging.
+
+## Testing changes locally (Windows)
+
+Follow all the steps above in Visual Studio and use Azure Storage Emulator for Windows to simulate the storage account.
```
Comment (ConnorMcMahon) [inline CONTRIBUTING.md]: I would clarify that the only steps above to do in Visual Studio is the Durable Extension .NET debugging. Even then, Windows JS developers may still feel more comfortable using VS Code, so maybe reword with something like the below:

Follow all the steps above, use the Azure Storage Emulator for windows to simulate the storage account, and optionally use Visual Studio to debug the .NET Durable Extension.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    - VS Code (or) Visual Studio
+- Tools
+    - [Azurite V2](https://github.com/Azure/Azurite/tree/legacy-master) (for MacOS) (or) [Azure Storage Emulator](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator) (or) Storage account in Azure
+    - [Azure Functions Core Tools](https://github.com/Azure/azure-functions-core-tools) v2.7.x and above.
+
+## Change flow
+
+The general flow for making a change to the script host is:
+1. 🍴 Fork the repo (add the fork via `git remote add me <clone url here>`
+2. 🌳 Create a branch for your change (generally use dev) (`git checkout -b my-change`)
```
[the remark this answers is not in the document]
Comment (priyaananthasankar) [reply CONTRIBUTING.md]: Done. https://github.com/Azure/azure-functions-durable-js/pull/124/commits/7f085c2fb23d722fb07afce41cf6f653b739a2ee

File: CONTRIBUTING.md (hunk 1 above)
Comment (priyaananthasankar) [reply CONTRIBUTING.md]: Azurite V2 (which is not the version officially recommended by Azurite) works. V3 does not support some storage mechanisms. Does it make sense to add to starter templates now or wait until Azurite stabilizes?

File: CONTRIBUTING.md (hunk 2 above)
Comment (priyaananthasankar) [reply CONTRIBUTING.md]: Done https://github.com/Azure/azure-functions-durable-js/pull/124/commits/7f085c2fb23d722fb07afce41cf6f653b739a2ee

**Comment by priyaananthasankar** [comment]
@scgbear : The primary instructions are to make changes in the JS library itself.

**Comment by priyaananthasankar** [comment]
@ConnorMcMahon : Sorry Got delayed reworking this, but here it is. Do let me know if it looks good to be merged.

File: CONTRIBUTING.md (hunk 1 above)
Comment (ConnorMcMahon) [reply CONTRIBUTING.md]: I see. It looks like there are no plans to support Tables in V3, due to Cosmos DB having a separate emulator for Azure Tables (which unfortunately, is NOT cross platform). You can see the [discussion here](https://github.com/Azure/Azurite/issues/253). With that in mind, maybe Azurite V2 is the best local dev option?

Comment (priyaananthasankar) [reply CONTRIBUTING.md]: @ConnorMcMahon : Yes given those limitations v2 is the best (as of now)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+## Pre-reqs
+
+- OS
+    - MacOS (or) Windows10
+- Language Runtimes
+    - .NET Core 2.0
+    - Python 3.6.x
+- Editor
+    - VS Code (or) Visual Studio
```
Comment (cgillum) [inline CONTRIBUTING.md]: Should we add links to the language runtime installation downloads and to the editors?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      "hubName": "{hubName}"
+    }
+  }
+}
+```
+
+- `func extensions install` (this will install an extensions.csproj that contains the version of DurableTask as seen below)
+
+```xml <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.DurableTask" Version="1.8.2" />
```
Comment (cgillum) [inline CONTRIBUTING.md]: I suggest we point people to 2.0.0 or greater.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    <PackageReference Include="Microsoft.Azure.WebJobs.Script.ExtensionsMetadataGenerator" Version="1.1.0" />
+  </ItemGroup>
+```
+
+### Setting up Durable Extension .NET debugging
+
+- Download the source code of DurableTask extension from [here](https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.DurableTask)
```
Comment (cgillum) [inline CONTRIBUTING.md]: Is it necessary to download the source code? We use SourceLink today and Visual Studio knows how to automatically download source code at debug time (does VS Code know how to do this?). I would hope that downloading the extension project would only be needed if they need to modify the extension code (super-advanced scenario).

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            "request": "attach",
+            "processId": "${command:pickProcess}",
+            "requireExactSource": false,
+            "justMyCode": false
+ }
+ ```
+
+### Debugging end-to-end
+
+1. Open the Azure Storage Explorer and connect to the local storage emulator or the storage account you are using.
```
Comment (cgillum) [inline CONTRIBUTING.md]: Should we add Azure Storage Explorer (with a download link) to the prereqs list above?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+2. In the VSCode editor for durable-js click Debug -> Start Debugging. This will internally start `func host start` through core tools and provides the orchestrator client URL
+3. In the VSCode editor for DurableTask, click Debug -> .NET Core Attach Process and search for `func host start` process and attach to it.
```
Comment (cgillum) [inline CONTRIBUTING.md]: Could this be considered a more advanced scenario, for people who are trying to debug both the extension and the JS library?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+4. Add a breakpoint in both editors and continue debugging.
+
+## Testing changes locally (Windows)
+
+Follow all the steps above, use the Azure Storage Emulator for windows to simulate the storage account, and optionally use Visual Studio to debug the .NET Durable Extension.
```
Comment (cgillum) [inline CONTRIBUTING.md]: In my experience, there is a mechanism for pointing your dev environment to get the sources from durable-functions npm package from you local enlistment rather than from npm.org. Should we add those instructions here?

File: CONTRIBUTING.md (hunk 4 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:16]: Good idea, done!

File: CONTRIBUTING.md (hunk 7 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:114]: There's already a link to it in the "Tools" section. I have updated the link to point to the download URL though

File: CONTRIBUTING.md (hunk 5 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:91]: I have updated the example `.csproj` and pointed out that we recommend sticking to versions equal or higher to `2.2.0`

File: CONTRIBUTING.md (hunk 8 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:116]: You are right, I have addressed it as such in my latest commit :)

File: CONTRIBUTING.md (hunk 6 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:98]: I'm reworking this section to make it clear that, if users want to debug the .NET side of things, they should use Visual Studio

File: CONTRIBUTING.md (hunk 9 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:115]: Added!

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  "version": "2.0",
+  "extensions": {
+    "durableTask": {
+      "hubName": "{hubName}"
+    }
+  }
+}
+```
+
+- `func extensions install` (this will install an extensions.csproj that contains the version of DurableTask as seen below). We recommend using version `2.2.0` or greater.
```
Comment (ConnorMcMahon) [inline CONTRIBUTING.md:88]: Instead of 2.2.0 or greater, why don't we just say "the latest version on nuget.org" and point to the package link.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+- C# for Visual Studio Code (powered by OmniSharp)
+- Azure Functions Extensions for Visual Studio Code v0.19.1 and above.
+
+### Setting up Durable JavaScript using sample code
+
+- Create a Durable Functions Orchestrator for FunctionChaining pattern using [starter templates](https://docs.microsoft.com/en-us/azure/azure-functions/durable/quickstart-js-vscode)
+  Note: In this starter template, ignore the line that says: "On a Mac or Linux computer, you must set the AzureWebJobsStorage property to the connection string of an existing Azure storage account". We will be setting up the AzureWebJobsStorage property to `UseDevelopmentStorage=true`
+
+- In host.json, remove the extensionsBundle portion to enable specific version debugging. Provide a hub name (else remove the entire extensions portion to default to DurableFunctionsHub) Here's how the host.json should look like:
```
Comment (ConnorMcMahon) [inline CONTRIBUTING.md:75]: This whole section about extension development could probably move down into the `Debugging end-to-end with the Durable Extension` section, as it technically isn't required if you are just doing development on our npm package.

File: CONTRIBUTING.md (hunk 9 above)
Comment (ConnorMcMahon) [reply CONTRIBUTING.md:115]: I think we should add that new chunk to this section, as opposed to the `Setting up Durable JavaScript using sample code` section.

File: CONTRIBUTING.md (hunk 10 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:88]: Addressed in next commit.

File: CONTRIBUTING.md (hunk 11 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:75]: Moved it :)

File: CONTRIBUTING.md (hunk 9 above)
Comment (davidmrdavid) [reply CONTRIBUTING.md:115]: resolved offline
